### ISSUE TITLE
Milestone: Version 0.1

### ISSUE THREAD

**Issue by quill18** [TeamPorcupine/ProjectPorcupine#605]
I have created our first milestone and, as a totally arbitrary date, set September 30th as the target date. No reason for it and I suspect we'll change it as we move forward.

I have flagged #49 as a milestone, because it seems like a pretty obvious "we need this".

Based on the votes, character needs - a simple first pass at least - should also be ready for version 0.1.  We should hammer down exactly what we expect from that. (Food, Oxygen, Rest should be a sufficient start to prove the system works.)

[![Feature Requests](http://feathub.com/TeamPorcupine/ProjectPorcupine?format=svg)](http://feathub.com/TeamPorcupine/ProjectPorcupine)

A power system of some kind seems to also be a basic expectation. Whether or not that includes power conduits for version 0.1 or just a global/automatic distribution (as is already partially in place) remains to be determined.  I think #602 and #604 are good discussions on various power topics.  I don't know that I'd worry about other utility systems yet (although air vents between adjacent rooms, at least, should be trivial).

**Within a couple days we should sort out what we feel is the minimum to have something that feels like a playable game** (and that probably means things like a main menu as well).

After that, we should officially split the project into three (?) branches:

- master is the stable/playable branch (though this is pretty much meaningless right now).
- devel is the current development branch for the next milestone. When it is "done", this will then be pulled into master.
- experimental is the branch for things that go above and beyond the next milestone.

People used to collaborating on open-source projects: Do these branches make sense, and do you have experience with syncing between them, say if we decide to pull a specific experimental feature into the devel branch?

**Comment by quill18** [comment]
P.S. You can sort Issues and Pull Requests by milestones.

**Comment by Mizipzor** [comment]
Awesome!

**Comment by RodRitter** [comment]
I can only speak for a company based system and those are the 3 branches we use. We also use Gitflow to organize our branching into Feature, Hotfix or Release folders. Not sure if that will be used but it does make things much more organized.

**Comment by mikejbrown** [comment]
I've never really worked on a big collaborative open source project, so am asking: as a general rule should new features be started on experimental first and merged (rebased?) into devel if and only if the roadmap gets updated saying the feature should go in the next milestone? That's not quite gitflow... but I'm happy to go with whatever @quill18 and the community wants of course.

**Comment by RodRitter** [comment]
I think once a release happens, Develop becomes Master and Experimental becomes Develop.

**Comment by mikejbrown** [comment]
I'm not sure if that should necessarily be automatic? I'd imagine experimental could have a mix of "almost ready for the next release" which would be fine and "pipe dream for waaay down the track" that you wouldn't want to automatically push onto devel (e.g. fully immersive 3D VR UI ;-)). Alternatively there could be a policy that nothing too far fetched gets onto TeamPorcupine/ProjectPorcupine/experimental, so e.g. VR work would only happen on somebodywhosePRswillnevergetaccepted/ProjectPorcupine/experimental.

**Comment by RodRitter** [comment]
The problem is that once something is merged into Experimental, how would we knitpick that specific feature and move it into Develop without digging it out of code and adding it?

**Comment by quill18** [comment]
I don't think experimental would ever become develop.  I think that after we pull develop into master, we then start to pull features for the next release from experimental into develop -- but never the entire experimental branch.

But I don't actually know how to cherry-pick features like that.

**Comment by Mizipzor** [comment]
eId like to point out that using forward slashes in branch names is common for creating structure. People wouldnt share the `experimental `branch, rather create `experimental/feature1`

**Comment by sboigelot** [comment]
@RodRitter you can cherry pick commit but... that's a pain

I'm a consultant in a big company with 200+ dev working on the same repositories. We have many git repo (per module) but most follow the same branch pattern:

- Devlop branch for the current developement
- Branch of develop for the release, the latest release branch is the stable version, if bug-fix need to be done on a specific version, it is applied on this branch first, then merged into develop
![image](https://cloud.githubusercontent.com/assets/6545998/17857849/a8d4e8ac-6883-11e6-9d09-952026fed1b4.png)
-Each feature is done on a branch of develop then merge when ready/peer reviewed and tested by qa

**Comment by mikejbrown** [comment]
@RodRitter I see what you're saying... I was thinking about where feature branches were based off of, not where things have already been merged in...

@quill18 So why not just follow gitflow here and base feature branches on devel? When a release happens you just merge in the features that are ready into devel then devel into master. I'm just confused about the use case for experimental here, i.e. when would you want to merge a feature onto it?

**Comment by bjubes** [comment]
Can we make devel just Dev? It sounds nicer IMO.

**Comment by WardBenjamin** [comment]
While I have seen some projects that have used `experimental` or `exp` branches pretty successfully, it's a pretty big recipe for lots of hardship if you have unexperienced contributors who don't really know what they're doing with git. I'd suggest keeping it simple with a stable `master` branch that only has critical hotfixes applied, and an unstable `develop` branch that has new feature development. Hotfixes would make it into `develop` if required (i.e. they haven't been superceeded already because of a redesign of something) through merging master into develop.

**Comment by Mizipzor** [comment]
Lets avoid shorthands. Say `experimental` not `exp`, and `develop` not `dev`. Its easier for everyone to read/understand. You dont type the names that often. Especially not if youre using a gui.

**Comment by sboigelot** [comment]
@sboigelot yep simplicity is the key. Keep it simple and stay out of complex system that will make everyone confused

**Comment by vogonistic** [comment]
I vote that instead of making our own system we adopt [Git Flow](https://github.com/nvie/gitflow). ([Git Flow Intro](http://nvie.com/posts/a-successful-git-branching-model/)) It has support in both SourceTree and GitKraken, as well as command line utilities. 

It uses `master` for offical releases, `develop` for merged features and when people work on additions they create a new `feature` branch that eventually gets merged into `develop`. When we are ready for a release, we merge `develop` into `master` and tag it. There are provisions for hotfixes that if accepted gets merged into both `master` and `develop`.

It's not a perfect system, but it's pretty easy to understand when you start working with it and has good support and documentation.

**Comment by WardBenjamin** [comment]
I like it. Seems... basically what I suggested :P

**Comment by vogonistic** [comment]
Both you @WardBenjamin and @RodRitter. I'm just repeating it to sound cool =)

**Comment by misspellted** [comment]
*pokes in*

Hmm, more branching discussions. Almost left this one alone, but figured I'd add nearly worthless thoughts on how I am expected to work at .. work.

We use GitFlow, maybe a bit modified? I'm not the process person at work, I just use it. It may be a bit of overkill if implemented here, but figured it may [probably not] be useful:

- `master` _The most recent commit is the most recent public release._ Created when the repository is started. Modified only via pull requests from `release/*` and `hotfix/*` branches. Merges generate `release/<$versionNumber>` tags.

- `release/<$versionNumber>` _Usually a pending release candidate, subject to testing and bug fixes._ Created when the boss decides enough implemented features warrant a release. Modified only via pull requests from `bugfix/` branches. Merged branches do not increment pending version number. At my job, a few release branches exist for legacy support reasons.

- `develop` _Summation of active development efforts._ Created immediately created after `master` by the repository creator. Modified directly only via pull requests from `feature/*` branches. Modified indirectly, if needed, via "cascading merges" automatically (if possible) from `hotfix/*` or `bugfix/*` branches.

- `hotfix/<Issue-Number_Short-Description>` _Oops! Shipped a bug and customer needs a fix yesterday!_ Branched from `master` by anyone on the development team. Merged into `master` via pull request, and via tools, can be automatically cascade-merged into `release/` and `develop` branches, if applicable. At my job, increments `$versionNumber` (only last group, so 3.0.0 -> 3.0.1).

- `bugfix/<Issue-Number_Short-Description>` _QA/Testing found a bug, boss wants it fixed before release._ Branched from `release/` by anyone on the development team. Merged into `release/` via pull request, and via tools, can be automatically cascade-merged into `develop` branch, if applicable. At my job, does not increment `$versionNumber`.

- `feature/<Issue-Number_Short-Description>` _A new feature (yay!), a bug fix that isn't schedule for immediate or a pending release, or general code clean up (style violations, unit test changes/additions/removals)._ Branched from `develop` by anyone on the development team. Merged into `develop` via pull request.

**Comment by vircung** [comment]
@vogonistic @WardBenjamin @RodRitter you're right. I would also go with git-flow approach but in limited scope. All we need is only :
- ```master``` (with tags?) for stable
- ```release/*``` for code freeze/hardening/bug fixing
- ```develop```  for unstable/WIP

Additionally every ```bugfix/```, ```hotfix/``` and ```feature/``` would be handled by PR to proper branch as we do now. No rocket science there.

**Comment by quill18** [comment]
I like it.  `release/*` and `develop` is clearer than `develop` and `experimental`.

**Comment by vogonistic** [comment]
We had a chat among the Gatekeepers (me, @WardBenjamin, @RodRitter and @Mizipzor) trying to figure out how to not break things when switching methods and avoid having to train all the new developers coming in that are interested in the project but doesn't have much git experience.

We came to the conclusion that it would avoid a lot effort if we keep the `master` branch as the Development branch. We'd go with `git-flow`, but if people use UI tools for it, they have to configure `master` as Development and `stable` as Master.

Branches:
- `master` for merged development against the next release
- `feature/*` for PRs targetting `master` and the next release
- `hotfix/*` for bugfixes in an already released version
- `release/*` for release specific branches, this is where the hotfixes go
- `stable` for the tagged releases and to get the latest stable version.

What does this mean? It means that development can continue as is today. We'll start managing things differently from our (Gatekeeper) side and we'd prefer that new PRs use the new `feature/` prefix, but that's it. 

The only really drawback is that people who use git flow needs to change the settings from the default and that `master` isn't as descriptive as `develop` when it comes to volatility.

@quill18 Are you on board?

**Comment by Mizipzor** [comment]
@vogonistic I'm going to assume that there will be questions, comments and concerns regarding this. Lets setup an issue dedicated to that so they dont end up here.

**Comment by vogonistic** [comment]
I created a separate issue for discussing the branches I mention here: #694

**Comment by quill18** [comment]
Closing this as many discussions have been split to their own issues.
